src/Juno/Runtime/Ledger.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE RecordWildCards #-}
+module Juno.Runtime.Ledger
+    (
+     lookupEntry,lastEntry,takeEntries
+    ,getEntriesAfter,lastLogInfo,logInfoForNextIndex
+    ,entryCount,maxIndex
+    ,appendLogEntry,addLogEntriesAt
+    ) where
+
+import Juno.Runtime.Types
+import Control.Lens
+import Data.ByteString (ByteString)
+import qualified Data.Sequence as Seq
+import Data.Sequence (Seq)
+import Codec.Digest.SHA
+import Data.Serialize
+
+-- | Get last entry.
+lastEntry :: Ledger a -> Maybe a
+lastEntry (_ :> e) = Just e
+lastEntry _ = Nothing
+
+-- | Get largest index in ledger.
+maxIndex :: Ledger a -> LogIndex
+maxIndex = subtract 1 . entryCount
+
+-- | Get count of entries in ledger.
+entryCount :: Ledger a -> LogIndex
+entryCount = fromIntegral . Seq.length . view lEntries
+
+-- | Safe index
+lookupEntry :: LogIndex -> Ledger LogEntry -> Maybe LogEntry
+lookupEntry i = firstOf (ix i)
+
+-- | take operation
+takeEntries :: LogIndex -> Ledger a -> Seq a
+takeEntries t = Seq.take (fromIntegral t) . _lEntries
+
+-- | called by leaders sending appendEntries.
+-- given a replica's nextIndex, get the index and term to send as
+-- prevLog(Index/Term)
+-- called by leaders sending appendEntries.
+-- given a replica's nextIndex, get the index and term to send as
+-- prevLog(Index/Term)
+logInfoForNextIndex :: Maybe LogIndex -> Ledger LogEntry -> (LogIndex,Term)
+logInfoForNextIndex mni es =
+  case mni of
+    Just ni -> let pli = ni - 1 in
+      case lookupEntry pli es of
+        Just LogEntry{..} -> (pli, _leTerm)
+         -- this shouldn't happen, because nextIndex - 1 should always be at
+         -- most our last entry
+        Nothing -> (startIndex, startTerm)
+    Nothing -> (startIndex, startTerm)
+
+
+-- | get the last term, index, hash of a log or default values if log empty.
+-- TODO really questionable that we're not just returning the last entry but
```

Comment:
tl;dr - just think of AER(-1,"") as our genesis block as it covers a bunch of edge cases on startup.

Yep, required actually to take care of a partition recovery edge case -- Leader keeps track of how long it's been since its heard from any follower (via AER). After 2x max(election timeouts) the leader steps down. As such, all followers need to be able to send AER's always unless we want some corner case logic.